# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/canvas/DrawTargetWebgl.cpp
# Commit: e51d895aa692
# Full Hash: e51d895aa6927c805cbb451ad6b299dc869d2e87
# Author: Lee Salzman <lsalzman@mozilla.com>
# Date: 2021-11-17 03:41:08
# Description:
#   Bug 1741453 - Unlink GlyphCacheEntry on failure. r=gfx-reviewers,jrmuizel
#   
#   If we fail to rasterize text for a GlyphCacheEntry, it may not have a linked TextureHandle by the time it is destructed.
#   Check that a handle exists before we try to unlink it. Also, if rasterization fails, just unlink it immediately so we
#   don't keep around a bunch of empty entries in the glyph cache.
# ==============================================================================

diff -r 4bedbdf30cea -r e51d895aa692 dom/canvas/DrawTargetWebgl.cpp
--- a/dom/canvas/DrawTargetWebgl.cpp	Tue Nov 16 21:23:29 2021 +0000
+++ b/dom/canvas/DrawTargetWebgl.cpp	Tue Nov 16 21:25:09 2021 +0000
@@ -181,6 +181,10 @@
 DrawTargetWebgl::DrawTargetWebgl() = default;
 
 DrawTargetWebgl::~DrawTargetWebgl() {
+  while (!mTextureHandles.isEmpty()) {
+    PruneTextureHandle(mTextureHandles.popLast());
+    --mNumTextureHandles;
+  }
   UnlinkSurfaceTextures();
   UnlinkGlyphCaches();
 }
@@ -403,7 +407,7 @@
   }
   if (!mVertexBuffer) {
     mVertexBuffer = mWebgl->CreateBuffer();
-    float rectData[8] = {0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 1.0f, 1.0f, 1.0f};
+    static const float rectData[8] = {0.0f, 0.0f, 1.0f, 0.0f, 0.0f, 1.0f, 1.0f, 1.0f};
     mWebgl->BindVertexArray(mVertexArray.get());
     mWebgl->BindBuffer(LOCAL_GL_ARRAY_BUFFER, mVertexBuffer.get());
     mWebgl->RawBufferData(LOCAL_GL_ARRAY_BUFFER,
@@ -1475,8 +1479,11 @@
   if (isInList()) {
     remove();
   }
-  mHandle->SetGlyphCacheEntry(nullptr);
-  mHandle = nullptr;
+  // The entry may not have a valid handle if rasterization failed.
+  if (mHandle) {
+    mHandle->SetGlyphCacheEntry(nullptr);
+    mHandle = nullptr;
+  }
 }
 
 GlyphCache::~GlyphCache() {
@@ -1639,6 +1646,9 @@
               // If drawing succeeded, then the text surface was uploaded to
               // a texture handle. Assign it to the glyph cache entry.
               entry->Link(handle);
+            } else {
+              // If drawing failed, remove the entry from the cache.
+              entry->Unlink();
             }
             return;
           }
